crypto/fipsmodule/kdf/sskdf.c

```diff
@@ -0,0 +1,314 @@
+    goto err;
+  }
+
+  md_ctx = EVP_MD_CTX_new();
+  if (!md_ctx) {
+    goto err;
+  }
+
+  ret = 1;
+  variant_ctx->digest = digest;
+  variant_ctx->ctx = md_ctx;
+  ctx->data = variant_ctx;
+
+  return ret;
+
+err:
+  EVP_MD_CTX_free(md_ctx);
+  OPENSSL_free(variant_ctx);
+
+  return ret;
+}
+
+
+
+static void SSKDF_VARIANT_DIGEST_CTX_cleanup(SSKDF_VARIANT_CTX *ctx) {
+  if (!ctx) {
+    return;
+  }
+  SSKDF_VARIANT_DIGEST_CTX *variant_ctx = (SSKDF_VARIANT_DIGEST_CTX *)ctx->data;
+  if (!ctx->data) {
+    return;
+  }
+  EVP_MD_CTX_free(variant_ctx->ctx);
+  OPENSSL_free(variant_ctx);
+  ctx->data = NULL;
+}
+
+static int SSKDF_VARIANT_HMAC_CTX_init(SSKDF_VARIANT_CTX *ctx,
+                                       const EVP_MD *digest,
+                                       const uint8_t *salt, size_t salt_len) {
+  SSKDF_VARIANT_HMAC_CTX *variant_ctx = NULL;
+  HMAC_CTX *hmac_ctx = NULL;
+
+  int ret = 0;
+
+  if (!ctx || !digest) {
+    goto err;
+  }
+
+  variant_ctx = OPENSSL_malloc(sizeof(SSKDF_VARIANT_HMAC_CTX));
+  if (!variant_ctx) {
+    goto err;
+  }
+
+  hmac_ctx = HMAC_CTX_new();
+  if (!hmac_ctx) {
+    goto err;
+  }
+
+  if (!HMAC_Init_ex(hmac_ctx, salt, salt_len, digest, NULL)) {
+    goto err;
+  }
+
+  ret = 1;
+  variant_ctx->ctx = hmac_ctx;
+  ctx->data = variant_ctx;
+
+  return ret;
+
+err:
+  HMAC_CTX_free(hmac_ctx);
+  OPENSSL_free(variant_ctx);
+
+  return ret;
+}
+
+
+
+static void SSKDF_VARIANT_HMAC_CTX_cleanup(SSKDF_VARIANT_CTX *ctx) {
+  if (!ctx) {
+    return;
+  }
+  SSKDF_VARIANT_HMAC_CTX *hmac_ctx = (SSKDF_VARIANT_HMAC_CTX *)ctx->data;
+  if (!hmac_ctx) {
+    return;
+  }
+  HMAC_CTX_free(hmac_ctx->ctx);
+  OPENSSL_free(hmac_ctx);
+  ctx->data = NULL;
+}
+
+static size_t sskdf_variant_digest_size(SSKDF_VARIANT_CTX *ctx) {
+  if (!ctx || !ctx->data) {
+    return 0;
+  }
+  SSKDF_VARIANT_DIGEST_CTX *variant_ctx = (SSKDF_VARIANT_DIGEST_CTX *)ctx->data;
+  return EVP_MD_size(variant_ctx->digest);
+}
+
+static int sskdf_variant_digest_compute(SSKDF_VARIANT_CTX *ctx, uint8_t *out,
+                                        size_t out_len,
+                                        const uint8_t counter[4],
+                                        const uint8_t *secret,
+                                        size_t secret_len, const uint8_t *info,
+                                        size_t info_len) {
+  if (!ctx || !ctx->data || !out || !counter || !secret) {
+    return 0;
+  }
+
+  SSKDF_VARIANT_DIGEST_CTX *variant_ctx = (SSKDF_VARIANT_DIGEST_CTX *)ctx->data;
+
+  if (!variant_ctx->ctx || !variant_ctx->digest) {
+    return 0;
+  }
+
+  if (!EVP_MD_CTX_reset(variant_ctx->ctx) ||
+      !EVP_DigestInit_ex(variant_ctx->ctx, variant_ctx->digest, NULL) ||
+      !EVP_DigestUpdate(variant_ctx->ctx, &counter[0], 4) ||
+      !EVP_DigestUpdate(variant_ctx->ctx, secret, secret_len) ||
+      !EVP_DigestUpdate(variant_ctx->ctx, info, info_len) ||
+      !EVP_DigestFinal(variant_ctx->ctx, out, NULL)) {
```

Comment:
should we check if `out_len` is large enough for `EVP_DigestFinal` to write its result to?

Reply:
I was borrowijg an optimization decision we take advantage of in `HKDF_expand` which is that we've sized `out` to be exactly `EVP_MAX_MD_SIZE` so we know the output will be no more then that value. We could technically add an assert here, but we also already check this exact condition in the `SSHKDF` function that orchestrates the specific compute variant functions. Here size outputs the size that `EVP_DigestFinal` or `HMAC_Final` will output:
```
size_t h_len = variant->size(ctx);
  if (!h_len) {
    goto err;
  }
  assert(h_len <= EVP_MAX_MD_SIZE);
```

So I think our bases are covered here already and additional checking would be redundant.